{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "rootf = FOLDER_NAME ## Put the folder path where folder RBM-MHC is saved ##\n",
    "\n",
    "## imports ##\n",
    "\n",
    "import sys, os, pickle\n",
    "sys.path.append(rootf + '/rbm/PGM3/source/')\n",
    "sys.path.append(rootf + '/rbm/PGM3/utilities/')\n",
    "from common_imports import *\n",
    "import dataset_utils, plots_utils, RBM_utils, evaluate_learning_utils, Proteins_utils, sequence_logo\n",
    "import utilities as utilities\n",
    "import rbm as rbm\n",
    "import importlib\n",
    "import butils"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We focus on sample 12T from Kalaora et al. 2016, Kalaora et al. 2018\n",
    "We first visualize motifs given by peptides of the same HLA binding specificity as \n",
    "predicted by RBM-MHC, as in Fig. 2a and Supplementary Fig. 8."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "clas = '12T'\n",
    "out_fold = rootf + '/rbm-mhc/data/Melanoma_samples/' + clas\n",
    "HLA_filename = out_fold + '/HLA-typing.txt'\n",
    "\n",
    "hlas = []\n",
    "with open(HLA_filename) as f:\n",
    "    for line in f:\n",
    "        linesplit = line.strip().split('\\n')\n",
    "        hlas.append(linesplit[0])\n",
    "nc = len(hlas)\n",
    "\n",
    "filename_motif = out_fold + '/full_sample_original_classification_YesReal_1_hu_10_l120.001_AL9.5_SA9_RWNoAlpha_TR1.txt'\n",
    "table1 = pd.read_csv(filename_motif , sep='\\t') ## extract data from original, patient-derived, dataset ##\n",
    "table2 = table1[table1['TR']==0]\n",
    "\n",
    "CC = 21 # number of symbols: amino acids + gap\n",
    "\n",
    "SA = 9 # length of the alignment ('sequence core')\n",
    "SAmin = 8 # minimal sequence length in the sample\n",
    "SAmax = 11 # maximal sequence length in the sample\n",
    "pseudocount = 0.000001\n",
    "\n",
    "import sequence_logo\n",
    "\n",
    "fig, axess = plt.subplots(6,1)\n",
    "fig.set_figheight(7)\n",
    "fig.set_figwidth(4)\n",
    "fig.subplots_adjust(wspace=0.5, hspace = 1.1, bottom = 0.2)\n",
    "st=12\n",
    "sl=st\n",
    "\n",
    "seqs_all=[]\n",
    "seqs_orig=[]\n",
    "\n",
    "for h in range(nc):\n",
    "    seqs_origh = list(table2[table2['RBM'] == h]['Data Core'].values)\n",
    "    seqs_allh = list(table1[table1['RBM'] == h]['Data Core'].values)\n",
    "    Nrbm=len(seqs_allh)\n",
    "    seqs_all.append(seqs_allh)\n",
    "    seqs_orig.append(seqs_origh)\n",
    "    sel_n = butils.convert_number(seqs_origh)\n",
    "    pwm_rbm = utilities.average(sel_n, c = CC);\n",
    "        \n",
    "    for i in range(SA):\n",
    "        for a in range(CC):\n",
    "            if pwm_rbm[i,a] < pseudocount:\n",
    "                pwm_rbm[i,a] = pseudocount  \n",
    "            \n",
    "    ax1= axess[h]\n",
    "    titlep =  hlas[h] + ' motif (# seqs. = ' + str(Nrbm) + ')'\n",
    "    sequence_logo.Sequence_logo(pwm_rbm, ax = ax1, figsize=(12,3.5), ylabel= 'bits', title = titlep, show=True, ticks_every=1, ticks_labels_size=sl, title_size=st)\n",
    "\n",
    "plt.tight_layout()\n",
    "namefig = out_fold + '/motifs_reconstructed.png'\n",
    "fig.savefig(namefig,dpi=300)\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we upload the trained RBM model, we visualize its weights and examples of inputs to hidden units, similarly to Fig. 1b and Supplementary Figs. 3-4."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "name_r = out_fold + '/model_YesReal_1_hu_10_l120.001_AL9.5_SA9_RWNoAlpha_TR1.data'\n",
    "RBM = RBM_utils.loadRBM(name_r)\n",
    "hu = RBM.n_h ## hidden units of the model ##\n",
    "\n",
    "## figure with weights ## \n",
    "fig, axess = plt.subplots(10,1)\n",
    "fig.set_figheight(11)\n",
    "fig.set_figwidth(4)\n",
    "fig.subplots_adjust(wspace=0.5, hspace = 1., bottom = 0.2)\n",
    "st=12\n",
    "sl=st\n",
    "\n",
    "for h in range(hu):           \n",
    "    ax1= axess[h]\n",
    "    titlep = ''\n",
    "    sequence_logo.Sequence_logo(RBM.weights[h], ax = ax1, figsize=(12,3.5), ylabel = r'W_' + str(h + 1), title=' ', ticks_every = 1,ticks_labels_size=sl,title_size=st)    \n",
    "plt.tight_layout()\n",
    "\n",
    "namefig = out_fold + '/weights.png'\n",
    "fig.savefig(namefig,  bbox_inches = 'tight', pad_inches = 0, dpi = 300)\n",
    "plt.close()\n",
    "\n",
    "colors = ['r','orange', 'deepskyblue', 'blue', 'springgreen', 'forestgreen']\n",
    "\n",
    "s2 = 18\n",
    "sl = 12\n",
    "list_ix=[2,6] ## example of hidden units, but any could be selected ##\n",
    "list_iy=[8] ## example of hidden units, but any could be selected ##\n",
    "s1 = 1.0\n",
    "\n",
    "for ix in list_ix:\n",
    "    for iy in list_iy:\n",
    "        fig, ax = plt.subplots()\n",
    "        fig.set_figwidth(5)\n",
    "        fig.set_figheight(5)\n",
    "\n",
    "        for h in range(nc):\n",
    "            sel_n = butils.convert_number(seqs_all[h])\n",
    "            In = RBM.vlayer.compute_output(sel_n, RBM.weights)\n",
    "            ax.scatter(In[:,ix],In[:,iy], c = colors[h], s = s1, label = hlas[h])\n",
    "\n",
    "        ax.set_xlabel('Input to h_' + str(ix+1), fontsize=s2)\n",
    "        ax.set_ylabel('Input to h_' + str(iy+1), fontsize=s2)\n",
    "        ax.tick_params(axis='both', which='major', labelsize = s2)\n",
    "        plt.tight_layout()\n",
    "        plt.legend(fontsize=10,markerscale=2,frameon=False, loc ='upper right')\n",
    "        pathfig = out_fold + '/Input_' + str(ix+1) + '_vs_Input_' + str(iy+1) + '.png'\n",
    "        fig.savefig(pathfig,  bbox_inches = 'tight', pad_inches = 0, dpi = 300)\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we plot inputs to hidden unit 10 which we found \n",
    "enables the discrimination of HLA-B51:01 binding modes, as shown in Fig. 2c."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tools for aminoacids\n",
    "aa = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V',  'W', 'Y','-']\n",
    "aadict_rev = {k:aa[k] for k in range(len(aa))}\n",
    "aadict = {aa[k]: k for k in range(len(aa))} \n",
    "\n",
    "cy = ['C']\n",
    "group_cy = [aadict[i] for i in cy]\n",
    "pol = ['Q','N','S','T']\n",
    "group_pol = [aadict[i] for i in pol]\n",
    "pos = ['K','R','H']\n",
    "group_pos = [aadict[i] for i in pos]\n",
    "neg = ['D','E']\n",
    "group_neg = [aadict[i] for i in neg]\n",
    "aro = ['F','W','Y'];\n",
    "group_aro = [aadict[i] for i in aro]\n",
    "ali = ['V','L','I','M'];\n",
    "group_ali = [aadict[i] for i in ali]\n",
    "tiny = ['A','P','G'];\n",
    "group_tiny = [aadict[i] for i in tiny]\n",
    "gap = ['-']\n",
    "group_gap = [aadict[i] for i in gap]\n",
    "\n",
    "hydrophobic = ['I', 'L', 'V', 'M', 'F', 'Y', 'W','P']\n",
    "hydrophilic = ['A', 'G', 'T', 'S', 'N', 'Q', 'D', 'E', 'R']\n",
    "group_hydrophilic = [aadict[i] for i in hydrophilic]\n",
    "group_hydrophobic = [aadict[i] for i in hydrophobic]\n",
    "\n",
    "prop = ['tiny','cy','neg','neg','aro','tiny','pos','ali','pos','ali','ali','pol','tiny','pol','pos','pol','pol','ali','aro','aro','gap']\n",
    "aadict_prop = {k:prop[k] for k in range(len(prop))}\n",
    "\n",
    "ind_hla = 3 # 'HLA-B*51:01'\n",
    "sel_n = butils.convert_number(seqs_orig[ind_hla])\n",
    "\n",
    "ind_1 = [] # Hydrophobic\n",
    "for i in range(len(sel_n)):\n",
    "    if sel_n[i][0] in group_hydrophobic and sel_n[i][1] in group_hydrophobic:\n",
    "        ind_1.append(i) # select sequences that have 'R' in position 3       \n",
    "ind_2 = [] # hydrophilic + charge\n",
    "for i in range(len(sel_n)):\n",
    "    if (aadict_prop[sel_n[i][0]]=='neg' or aadict_prop[sel_n[i][0]]=='pol') and sel_n[i][1] in group_hydrophilic:\n",
    "        ind_2.append(i)   \n",
    "\n",
    "fig, ax1 = plt.subplots()\n",
    "fig.set_figwidth(5)\n",
    "fig.set_figheight(5)\n",
    "ind_h = 9 # 'HLA-B*51:01' \n",
    "\n",
    "In = RBM.vlayer.compute_output(sel_n, RBM.weights)\n",
    "\n",
    "ax1.hist(In[ind_1,ind_h],weights=None, bins=100, color = 'DimGrey', label = 'Hydrophilic/charge')\n",
    "ax1.hist(In[ind_2,ind_h],weights=None, bins=100, color = 'purple', label = 'Hydrophobic')\n",
    "plt.legend(fontsize=sl,markerscale=4,frameon=False, loc = 'upper left')\n",
    "ax1.tick_params(axis='both', which='major',labelsize = sl+2)\n",
    "ax1.set_xlabel('Input to h_' + str(ind_h + 1),fontsize=sl+2)\n",
    "ax1.set_ylabel('Count',fontsize=sl+2)\n",
    "pathfig = out_fold + '/hist_Input_' + str(ind_h+1) + '.png'\n",
    "fig.savefig(pathfig)\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Ranking neoantigens**:\n",
    "Here, we read the tables with scores assigned to mutant peptides in the patient's WES and we find the rank associated to MS-validated neoantigens, as in Supplementary Table 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/barbara/.local/lib/python3.6/site-packages/ipykernel_launcher.py:104: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.\n"
     ]
    }
   ],
   "source": [
    "### Read different tables ##\n",
    "out_fold_NA = out_fold + '/WESranking' \n",
    "na_na = ['DANSFLQSV','GVYPMPGTQK']\n",
    "na_hla = ['HLA-B*51:01','HLA-A*03:01']\n",
    "\n",
    "fn = out_fold_NA + '/WES_YesReal_1_hu_10_l120.001_AL9.5_19_SA9_RWNoAlpha_TR1.txt'\n",
    "wes_out = pd.read_csv(fn, sep = '\\t')\n",
    "fn = out_fold_NA + '/WES_YesReal_1_hu_10_l120.001_AL9.5_19_SA9_RWNoAlpha_TR1_pred.txt'\n",
    "predvec = np.loadtxt(fn)\n",
    "fnkm = out_fold_NA + '/WES_YesReal_1_hu_10_l120.001_AL9.5_19_SA9_RWNoAlpha_TR1_pred_kmeans.txt'\n",
    "predveckm = np.loadtxt(fnkm)\n",
    "\n",
    "wespeps = list((wes_out['wes']).values)\n",
    "wespeps_g = list((wes_out['wes_g']).values)\n",
    "likwes = list((wes_out['LL']).values)\n",
    "lenwes = list((wes_out['Len']).values)\n",
    "predwes = list((wes_out['pred']).values)\n",
    "predwesl = list((wes_out['pred_label']).values)\n",
    "score_mixmhc = list((wes_out['score_mixmhc']).values)\n",
    "val_score_mixmhc = list((wes_out['val_score_mixmhc']).values)\n",
    "val_score_pwmrbm = list((wes_out['val_score_pwmrbm']).values)\n",
    "score_kmeans = list((wes_out['labels_kmeans']).values)\n",
    "val_score_kmeans = list((wes_out['val_score_kmeans']).values)\n",
    "\n",
    "ent = []\n",
    "for t in range(len(predvec)):\n",
    "    ent.append(-sum(predvec[t]*np.log(predvec[t]+0.000001)))\n",
    "\n",
    "entkm = []\n",
    "for t in range(len(predveckm)):\n",
    "    entkm.append(-sum(predveckm[t]*np.log(predveckm[t]+0.000001)))\n",
    "\n",
    "## define a presentation score ##\n",
    "likwes_f = likwes - np.array(ent)\n",
    "kmeans = likwes - np.array(entkm)\n",
    "\n",
    "NAlist = zip(wespeps, wespeps_g, likwes_f, lenwes, predwes)\n",
    "ind_ll = 2\n",
    "ind_len = 3\n",
    "NArankL = sorted(sorted(NAlist, key = lambda x : x[ind_len]), key = lambda x : x[ind_ll], reverse = True)\n",
    "rankedwes = list(np.array(NArankL)[:,0])\n",
    "\n",
    "rankll=[]\n",
    "for t in range(len(rankedwes)):\n",
    "    rankll.append(wespeps.index(rankedwes[t]))\n",
    "\n",
    "NAlist = zip(wespeps, wespeps_g, val_score_mixmhc, lenwes, predwes)\n",
    "NArankM = sorted(sorted(NAlist, key = lambda x : x[ind_len]), key = lambda x : x[ind_ll], reverse = True)\n",
    "rankedwesm = list(np.array(NArankM)[:,0])\n",
    "rankmm=[]\n",
    "for t in range(len(rankedwesm)):\n",
    "    rankmm.append(wespeps.index(rankedwesm[t]))\n",
    "\n",
    "NAlist = zip(wespeps, wespeps_g, kmeans, lenwes, predwes)\n",
    "NArankK = sorted(sorted(NAlist, key = lambda x : x[ind_len]), key = lambda x : x[ind_ll], reverse = True)\n",
    "rankedwesk = list(np.array(NArankK)[:,0])\n",
    "rankkm=[]\n",
    "for t in range(len(rankedwesk)):\n",
    "    rankkm.append(wespeps.index(rankedwesk[t]))\n",
    "\n",
    "## perform ranking for RBM-MHC and RBM-km ##\n",
    "lna=[]\n",
    "lhla=[]\n",
    "lrbm_hla=[]\n",
    "lrbm_lik=[]\n",
    "lrbm_rank=[]\n",
    "lkm_hla=[]\n",
    "lkm_score=[]\n",
    "lkm_rank=[]\n",
    "km_score=[]\n",
    "rbm_score=[]\n",
    "for t in range(len(na_na)):\n",
    "    if na_na[t] in wespeps:  \n",
    "        ind = wespeps.index(na_na[t])\n",
    "        lna.append(na_na[t])\n",
    "        lhla.append(na_hla[t])\n",
    "        # pred RBM\n",
    "        lrbm_hla.append(na_hla[t] == hlas[predwesl[ind]])          \n",
    "        lkm_hla.append(na_hla[t] == hlas[score_kmeans[ind]])\n",
    "        lkm_score.append(format(val_score_kmeans[ind], '.3f'))\n",
    "        lrbm_lik.append(format(likwes[ind]/np.log(10), '.3f'))\n",
    "        km_score.append(likwes_f[ind])\n",
    "        rbm_score.append(kmeans[ind])\n",
    "        lrbm_rank.append(list(np.array(wespeps)[rankll]).index(na[t])+1)\n",
    "        lkm_rank.append(list(np.array(wespeps)[rankkm]).index(na[t])+1)\n",
    "\n",
    "title_text = 'Neoantigen prediction 12T (total candidates = ' + str(len(wespeps))+ ')'\n",
    "labels_rank = ['Neoantigen', 'Rank RBM-MHC', 'HLA-I predicted?', 'Rank RBM-kM', 'HLA-I predicted?' ]\n",
    "values = np.array([lna,lrbm_rank, lrbm_hla,lkm_rank, lkm_hla,])\n",
    "\n",
    "# Build a dataframe with ranking #\n",
    "\n",
    "from pandas.plotting import table\n",
    "\n",
    "df = pd.DataFrame(np.transpose(values), columns=labels_rank)\n",
    "fig, ax = plt.subplots(figsize = (2*len(values),len(na_na))) # set size frame\n",
    "ax.xaxis.set_visible(False)  # hide the x axis\n",
    "ax.yaxis.set_visible(False)  # hide the y axis\n",
    "ax.set_frame_on(False)  # no visible frame, uncomment if size is ok\n",
    "tabla = table(ax, df, loc='center')  # where df is your data frame\n",
    "tabla.auto_set_font_size(False) # Activate set fontsize manually\n",
    "tabla.set_fontsize(12)\n",
    "plt.suptitle(title_text, fontsize = 15)\n",
    "fig.show()\n",
    "pathfig = out_fold_NA + '/TableNArank.pdf'\n",
    "fig.savefig(pathfig)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
